Add ha to monasca barclamp

samirjorina opened the pull request.

jgrassler committed:
Removed monasca-api role.

jgrassler committed:
Added up-to-date WSGI plumbing.

jgrassler committed:
Addressed Hound comments.

samirjorina committed:
Install monasca-ui horizon plugin

samirjorina committed:
Fix barbican schema template

Commit:
Initial commit for the log-agent

Following is the first one that will
add installing monasca-log-agent
capabilities

kornicameister committed:
Use dedicated users for monasca-log-agent (#11)

* Use deticate users for monasca-log-agent

Following commit provides registering custom
project and dedicated user registered inside
this project

* Added some debug details

* Added including monasca:common

* Include monasca:common elsewhere

* Removed return clause'

* Removed action:nothing

* Move log outside block and added some variables

* Added newline

* Removed block

* Added new configuration variables into conf template

* TNT

* Use group instead of 2xowner

* Removed extra " from log-agent.conf.erb

* Changed way to access values

* Added user&group into databad

* Moved registering log-agent

Since agents (metric&logs) will be configured
with single credentials per each node it made sense
to put creating them inside the common part.

Additionally added randomizing the agents
password

* Fixing hound error

* Fixing hound error

* Fixed typo

* Addressed matrixik comments

* Fixing undefined method in monasca_service.rb

* Use empty strings instead of null values

* trying to make it work

* trying to make it work 2

* trying to make it work 3

* trying to make it work 4

* trying to make it work 5

* trying to make it work 6

* polished

* Pleasing Hound

samirjorina committed:
Configure monasca-ui plugin (#6)

* Configure monasca-ui plugin

* Refactor the code

* Addresing review comments

* Deploy grafana alongside horizon

* Remove obsolete changes to monasca schema

* Removing extra blank line

samirjorina committed:
Add ha to monasca barclamp

samirjorina committed:
Add monitoring dashboards to Grafana

* Add monitoring dashboards to Grafana

* Change template to cookbook_file

samirjorina committed:
Provide Ansible integration

* Provide Ansible integration - draft version

* Adding parameters to UI

* Adding group_all template

* Fixing inventory file templating

* Making sure Ansible runs successfully (at least once)

* Adapt constraints for monasca-server

* Set default deployment of monasca roles

* Add config field for CMM installer URL

* Fixing invalid merge

* Fix order of roles

* Add grafana database password

* Fixing hound violations

* Fixing hound violations 2

* Draft of inventory file template for cluster

* Adjust monasca-master role to use appropriate inventory file template

* Fill in inventory file for cluster

* Provide Ansible integration - draft version

* Adding parameters to UI

* Adding group_all template

* Fixing inventory file templating

* Making sure Ansible runs successfully (at least once)

* Adapt constraints for monasca-server

* Set default deployment of monasca roles

* Add config field for CMM installer URL

* Fixing invalid merge

* Fix order of roles

* Add grafana database password

* Fixing hound violations

* Fixing hound violations 2

* Addressing review comments

* Fix Hound violations

* Fix Hound violations 2

* Fix Hound violations 3

* Addressing review comments

* Addressing some additional comments

* Addressing some additional comments 2

samirjorina committed:
Add missing end keyword (#16)

* Add missing end keyword

* Fix indent

matrixik committed:
Fix json indentation (#17)

matrixik committed:
Update monasca-agent cookbook and barclamp (#2)

* Add monasca-reconfigure and sudoers file

* Fix monasca-setup bin path

* Fix template variables

* Keystone settings for agent

* Keystone settings for agent

* Keystone settings for agent

* Add log

* More readable config templates

* Remove log

* Remove per-node credentials todos

* Fix json indentations

* Rename agent dimensions variable name

* Run monasca-reconfigure every time with chef client

* Add firts GUI elements

* Add GUI elements

* Update for new code arrangement for keystone

* Fix yaml for translations

* Fix monasca helper module name

* Use root user for agent temporally

* Update monasca agent GUI parts

* WIP: Add custom checks for monasca-agent

* Extract variables for templates

* Fix testing filepath

* Fix template for monasca agent checks

* Fix template paths again

* Address Hound warning

* Add more monasca agent checkers

* Fix some variables

* Add log api port to data_bag

* Build Kibana and log-api urls

* Add more info to GUI

* Add user_managed: false

* Fix yaml indentation

* Add some logs

* Fix monasca-reconfigure

* Fix ruby boolean name

* Uncomment agent.yaml template code

* Fix providing service variable to template

* Fix insecure args for monasca-setup

* Fix monasca-reconfigure

* Configure more variables with monasca-reconfigure

* Add more config to UI

* Disable generating agent.yaml

It's regenerated every time chef client is run, it's running
monasca-reconfigure.

* Change default overwrite config to false

* Add monasca-agent to sudoers

* Run monasca-setup as monasca-agent user

* Remove logs from monasca-reconfigure

* Rename monasca-agent to monasca-metric-agent

And rename all variables to reflect this change.

* Fix hound warnings

* Address comments

* Fix copyright

* Use same order in data_bag schema like in json

* Restore user mapping in log_agent schema

* Change monasca-master unique to false

* Fix monasca-metric-agents names in json

* Use keystone settings from data_bag for metric agent

* Replace `owner` with `user` in execute command

* Replace group `monasca-agent` with `monasca`

* Remove adding monasca-agent to sudoers

Will be added from rpm.

* Add port to monasca_api_url for monasca-agent

* Use helper for getting api url

* Fix bad default config name

Commit:
Disable systemd service file creation (#19)

Enforce skip_enable option for monasca_setup

'skip_enable' option controls the creation of systemd service file for
Monasca Agent. The service file is created in RPM package, so the option
has to be enabled here.

samirjorina committed:
Fix invalid files location

matrixik committed:
Get proper log-api url for monasca-log-agent (#18)

* Get proper log-api url for monasca-log-agent

* Remove useless monasca urls from data_bag

* Remove log_agent user and group from data_bag

* Always use http in log-api url

* Use only http for monasca-agent url

* Remove ability to set `insecure` and `ca_file` options from UI.

samirjorina committed:
Finish integration of Ansible installer (#15)

* Finish integration of Ansible installer

* Set credentials in command line

* Fix violations

* Fix violations 2

* Add support for multiple networks in single node

* Pass missing variables to ansible

* Fix hound violations

* Use monasca.yml as entry point for ansible installer

* Fix paths and influxdb url

* Remove agent nodes from inventory file

matrixik committed:
Use sudo for running monasca-reconfigure (#21)

dirkmueller commented on chef/cookbooks/monasca/attributes/default.rb:
```diff
@@ -41,3 +41,8 @@
+
+default[:monasca][:ha][:enabled] = false
+# Ports to bind to when haproxy is used for the real ports
+default[:monasca][:ha][:ports][:plain] = 5580
```
thats conflicting with horizon, choose a different port

samirjorina replied:
This is just a 'random' port, I only wanted to experiment with HA to make sure I get it right. So can I assume that, in overall, the approach itself is correct?

jgrassler replied:
Yes, but please find a port that doesn't clash with anything else. As far as I know we do not have a fixed convention for that. Please correct me if I'm wrong about that, Dirk.

matrixik replied:
Done

dirkmueller commented on chef/cookbooks/monasca/recipes/ha.rb:
```diff
@@ -0,0 +1,26 @@
+
+haproxy_loadbalancer "monasca" do
+  address "0.0.0.0"
+  port 1234
```
is this the proper port?

samirjorina replied:
Please see above

jgrassler replied:
```
haproxy_loadbalancer "monasca-api" do
  address network_settings 0.0.0.0 # not sure if/how we can specify multiple addresses for this resource but this will make it listen on both the admin and public interface
  port 8070
  servers CrowbarPacemakerHelper.haproxy_servers_for_service(node, "monasca", "monasca-server", "monasca-api")
  )
  action :nothing
end.run_action(:create)

haproxy_loadbalancer "monasca-log-api" do
  address network_settings 0.0.0.0 # not sure if/how we can specify multiple addresses for this resource but this will make it listen on both the admin and public interface
  port 5607
  servers CrowbarPacemakerHelper.haproxy_servers_for_service(node, "monasca", "monasca-server", "monasca-log-api")
  )
  action :nothing
end.run_action(:create)

haproxy_loadbalancer "monasca-logs-search" do
  address network_settings 0.0.0.0 # not sure if/how we can specify multiple addresses for this resource but this will make it listen on both the admin and public interface
  port 5601
  servers CrowbarPacemakerHelper.haproxy_servers_for_service(node, "monasca", "monasca-server", "monasca-logs-search")
  )
  action :nothing
end.run_action(:create)
```

For this to work you'll need to define the ports where the regular services listen in the HA case in defaults.rb:

```
default[:monasca][:ha][:ports][:monasca-api] = 18070
default[:monasca][:ha][:ports][:monasca-log-api] = 15607
default[:monasca][:ha][:ports][:monasca-logs-search] = 15601
```

This is where `haproxy_servers_for_service` is defined:

https://github.com/crowbar/crowbar-ha/blob/master/chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb#L244

That should make it clear why this works out. It _is_ a bit convoluted, I'm afraid.

Also, I just used the regular listen port for the `haproxy_loadbalancer`
resource's port attribute above for the sake of clarity. Please don't do this
in your cookbook. Use a helper the way the [Barbican 
barclamp](https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/barbican/recipes/ha.rb#L23)
does it. You'll find it defined here:

https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/barbican/libraries/helpers.rb

That helper ensures the actual services listen on the alternate ports (i.e.
18070, 15607, 15601 in our case) in the HA case, and on the regular ports 
(8070, 5607, 5601 in our case) otherwise. To that end it provides the network
settings in both the [service recipe](https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/barbican/recipes/api.rb#L37)
and the [HA recipe](https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/barbican/recipes/api.rb#L37).

matrixik replied:
Done

jgrassler reviewed (changes requested):
Not quite there, yet unfortunately :-(

See my inline comments, especially the second one. I hope that makes things somewhat clearer (our lookup mechanisms for HA/non-HA ports _are_ a bit convoluted). Please feel free to ask if it's not clear (I took a while to wrap my mind around all these twists and turns myself the first time I added HA support to a barclamp...)

samirjorina committed:
Remove default grafana config file (#22)

matrixik committed:
Fix monasca-log-agent rpm name (#24)

samirjorina committed:
Fix suse (#23)

* Register monasca endpoints in Keystone

* Fix kibana url

* Rearrange helper functions

artur-ba committed:
Remove unused group (#27)

Remove unused grafana_group from cluster template

matrixik committed:
Disable deprecation_warnings for Ansible (#34)

They are outputted to STDERR and break applying cookbook.

samirjorina committed:
Agent log dir (#33)

* Register monasca endpoints in Keystone

* Fix kibana url

* Rearrange helper functions

* Make agent log dir configurable

samirjorina committed:
Configure agent service name for monasca setup (#32)

Override default monasca-agent service name with 
--agent_service_name flag

samirjorina committed:
Address review comments (#35)

samirjorina committed:
Fix agent config (#31)

Fix various issues

Fix grafana config

Fix memcached_nodes config

Fix memcached_nodes config 2

Remove commented grafana config

matrixik committed:
Update HA for monasca barclamp with all services

matrixik committed:
Uncomment code for cluster validation

matrixik committed:
Address Hound comments

artur-ba committed:
Remove offline resources (#36)

Remove all variables connected to offline resources.

matrixik commented:
PTAL

artur-ba commented on chef/cookbooks/monasca/libraries/helper.rb:
```diff
@@ -14,62 +14,121 @@
+
+        api: {
+          bind_port: bind_port_api,
+          ha_bind_host: node[:monasca][:api][:bind_host] == "*" ? "0.0.0.0" : @cluster_admin_ip,
```
I think it should be @cluster_internal_ip. The same comment for rest of the components in this file.

matrixik replied:
Renamed to `@cluster_monitoring_ip`

artur-ba commented on chef/cookbooks/monasca/attributes/default.rb:
```diff
@@ -52,3 +49,19 @@
+default[:monasca][:ha][:ports][:kibana] = 15601
+# Internal network
+default[:monasca][:ha][:ports][:mariadb] = 13306
+default[:monasca][:ha][:ports][:influxdb] = 18086
```
Please remember in the future to change this port to InfluxDB Relay port

matrixik replied:
Added InfluxDB Relay port but need to learn how to use `frontend` and `backend` HAProxy configuration with barclamps. For now added `FIXME` comments in code.

matrixik committed:
Add influxdb relay config

Rename cluster_admin_ip to cluster_monitoring_ip

matrixik committed:
Address hound comments

matrixik committed:
Add frontend for InfluxDB

matrixik committed:
Fix getting port for api and log-api servers

matrixik committed:
Fix loadbalancer for influx parts

matrixik committed:
Remove including HA from role_monasca_master

matrixik committed:
Run monasca-reconfigure only when the file changes (#38)

monasca-reconfigure detection script can run for a long time, so run it
only when the file is changed by Chef. As of now user will need to run
this script by hand to detect new services if the file is not changed.

matrixik committed:
Add more config options for HAProxy

matrixik committed:
Remove unnecessary configuration

Also set mode "tcp" for all InfluxDB configs.

matrixik committed:
Update hosts-cluster with proper virtual IP

matrixik committed:
Replace all cmm occurrences with monasca (#37)

* Replace all cmm occurrences with monasca

* Update monasca_service.rb

* Update master.rb

* Update all_group.erb

* Update template-monasca.json

* Update template-monasca.schema

artur-ba committed:
Use node_search_with_cache method (#39)

Use node_search_with_cache() method insted of search().
This will work better on nodes that do not have monasca-server
role applied.

matrixik committed:
Use only admin and public networks

samirjorina committed:
Log agent config (#40)

* Add config for log agent

* Allow manual config of log agent

* Remove log agent config from ui and update template

* Fix hound violations

* Fix hound violations 2

samirjorina committed:
Start and enable agents after installation

matrixik committed:
Use different ports for services running behind HAProxy

matrixik committed:
Use proper influxdb ports in cluster

matrixik committed:
Remove haproxy and keepalived hosts from hosts-cluster

matrixik committed:
Add more helpers

matrixik committed:
Fix setting proper ports in cluster
